EIPS/eip-3675.md

```diff
@@ -0,0 +1,247 @@
+
+This document specifies the set of changes to the block structure, block processing, fork choice rule and network interface introduced by the consensus upgrade.
+
+
+## Motivation
+
+The beacon chain network has been up and running since December 2020. Neither safety nor liveness failures were detected during this period of time. This long period of running without failure demonstrates the sustainability of the beacon chain system and its readiness to become a security provider for the Ethereum Mainnet.
+
+To understand the motivation of introducing the Proof-of-Stake consensus see the Motivation section of [EIP-2982](./eip-2982.md#motivation).
+
+
+## Specification
+
+### Definitions
+
+* **PoW block**: Block that is built and verified by the existing proof-of-work mechanism. In other words, a block of the Ethereum network before the consensus upgrade.
+* **PoS block**: Block that is built and verified by the new proof-of-stake mechanism.
+* **Terminal PoW block**: A PoW block that satisfies the following conditions -- 
+`pow_block.total_difficulty >= TERMINAL_TOTAL_DIFFICULTY` *and* `pow_block.parent_block.total_difficulty < TERMINAL_TOTAL_DIFFICULTY`.  
+There can be more than one terminal PoW block in the network, e.g. multiple children of the same pre-terminal block.
+* **`TERMINAL_TOTAL_DIFFICULTY`** The amount of total difficulty reached by the network that triggers the consensus upgrade.
+* **`TRANSITION_BLOCK`** The earliest PoS block of the canonical chain, i.e. the PoS block with the lowest block height.  
+`TRANSITION_BLOCK` **MUST** be a child of a terminal PoW block.
+* **`POS_CONSENSUS_VALIDATED`** An event occurring when the PoS block is validated with respect to the proof-of-stake consensus rules.
+* **`POS_CHAINHEAD_SET`** An event occurring when the PoS block is designated as the head of the canonical chain.
+* **`POS_BLOCK_FINALIZED`** An event occurring when the PoS block is considered as finalized.
+
+#### PoS events
+
+Events having the `POS_` prefix in the name (PoS events) are emitted by the new proof-of-stake consensus mechanism. They signify the corresponding assertion that has been made regarding a block specified by the event. The underlying logic of PoS events can be found in the beacon chain specification. On the occurrence of each PoS event the corresponding action that is specified by this EIP **MUST** be taken.
+
+Each `POS_BLOCK_FINALIZED` event has a reference to the head of the canonical chain in addition to the finalized block and incorporates the semantics of the `POS_CHAINHEAD_SET` event. Thus, every action and effect that is specified on the occurrence of `POS_CHAINHEAD_SET` event according to this EIP **MUST** also be taken on every occurrence of `POS_BLOCK_FINALIZED` event.
+
+
+### PoW block processing
+
+PoW blocks that are descendants of any terminal PoW block **MUST NOT** be imported. This implies that a terminal PoW block will be the last PoW block in the canonical chain.
```

Comment:
You mean that there is a bit of inconsistency between the two sections? Like, this section says keep processing terminal PoW blocks and their ancestors and puts no time boundary on it while the network section implies more strict behaviour and deprecates gossip at a strict boundary. And we might want to reflect this restriction here as well?

Reply:
I'd suggest to add the following kind of statement:
> Beginning with the first `POS_BLOCK_FINALIZED` event, client software **MUST NOT** import _any_ PoW block that doesn't belong to the canonical chain.

Reply:
My concern is that currently I don't think clients gossip siblings until they see a new longest chain (maybe I'm wrong about this)?  If that assumption is true, then this will be a behavior change that clients need to implement.

Reply:
Hmm, interesting. I would assume they do verify PoW upon receiving a block and propagate this block further if the verification passed. In the mean time this block is stored. And here we have options, it either processed right away or processing is deferred until the fork that this block does belong to becomes the longest chain.

And this is where we may have a real issue (not related to gossip). If the terminal PoW block that is picked by the beacon chain proposer has been received but hasn't been processed by our node due to conditions described above (our node may receive another terminal block with the same TD and thus refuse to re-org). Then our node won't be able to verify the beacon block that has been received because it will stuck verifying the PoW block that is the parent of the execution payload. So, it means that the spec must force _every_ terminal PoW block to be processed, not only stored. 

cc @holiman

Reply:
An alternative solution for this edge case would be the corresponding method in the Consensus API, let's say `verifyTerminalPoWBlock(block_hash: Hash32)`. This method would enforce the execution of a sibling terminal PoW block if it hasn't been executed yet. The downside of this approach is the delay between receiving a beacon block and considering it for further processing due to calling to this method, especially, if there are more than one block on this sibling fork that must be executed. But this approach seems to be less intrusive in terms of mainnet clients change set.

Reply:
A solution with `verifyTerminalPoWBlock(block_hash: Hash32)` or a similar call aids mitigation of the DoS vector that is described in https://github.com/ethereum/EIPs/pull/3675#discussion_r679207367. If miners have a cheap way to produce terminal PoW blocks to spam the network it is better for a node to avoid executing them all and execute terminal PoW blocks from non-canonical chains on demand.